---
## Hausaufgabe 1 - Grundlagen Python & Objektorientierung
---

### Achtung: Überprüfen Sie vor Abgabe der Hausaufgabe, ob das Notebook richtig gespeichert wurde. 
Die Speicherung von Notebooks funktioniert über das "Disketten"-Symbol im Notebook oder über die Shortcuts Strg + S (Win) und CMD + S (MAC).

---
## Bewertung
---

#### Erreichbare Punkte: 25
---
#### Erreichte Punkte:   24,5
---
---
Aufgabe 1: 6 <br>
<br>
Aufgabe 2: 6,5 <br>
<br>
Aufgabe 3: 12 <br>

### Aufgabe 1 - Listen & Sortierungen (6 Punkte)

Ihnen ist die folgende Liste gegeben.


In [1]:
thisList = ['Otto', 'Level', 'Leseesel', [1, 2, 3], 'Kajak', 'Mueller', 'Retter', 'Rotor', [5, 6, 7], 'Effe', 'Meier']

#### a.) 
#### (2 Punkte)

Erzeugen Sie eine zweite Liste *thisList2*, die die Einträge von *thisList* übernimmt. Jedoch sollen alle Elemente in der neuen Liste umgekehrt sein.
Informieren Sie sich dazu über die Möglichkeiten des Index-Operators und nutzen Sie diesen, um die gewünschte Ausgabe zu erzeugen.

In [2]:
### Ergänzen Sie hier Ihren Code, um die gewünschte Ausgabe zu erzielen. ###
thisList2 = []
for item in thisList:
    length = len(item) - 1
    if type(item) == str:
        item2 = ''
        while length >= 0:
            item2 += item[length]
            length -= 1
    else:
        item2 = []
        while length >= 0:
            item2.append(item[length])
            length -= 1
    thisList2.append(item2)




### Führen Sie anschließend nachfolgenden Programmcode aus, um Ihre Lösung zu überprüfen ###

print(thisList2)

## Richtig: 2 Punkte
## Sehr umständlich gelöst. Gleiches Ergebnis wird erzielt durch thisList2 = [x[::-1] for x in thisList]

['ottO', 'leveL', 'leseeseL', [3, 2, 1], 'kajaK', 'relleuM', 'retteR', 'rotoR', [7, 6, 5], 'effE', 'reieM']


In [3]:
thisList = ['Otto', 'Level', 'Leseesel', [1, 2, 3], 'Kajak', 'Mueller', 'Retter', 'Rotor', [5, 6, 7], 'Effe', 'Meier']


thisList2 = [x[::-1] for x in thisList]
print(thisList2)


['ottO', 'leveL', 'leseeseL', [3, 2, 1], 'kajaK', 'relleuM', 'retteR', 'rotoR', [7, 6, 5], 'effE', 'reieM']


In [8]:
a = [0, 1, 2, 3, 4]

print(a[0:5])
print(a[0: ])
print(a[0:5:2])

[0, 1, 2, 3, 4]
[0, 1, 2, 3, 4]
[0, 2, 4]


**解析*
一般情况下，a[i: j: s] 表示取列表a的第i项到第j-1项，按照步长为s。表示列表的中括号中可以有两个冒号，使用的时候注意冒号的位置。


步长s的取值分正负两种情况：
当s为正数时：把列表从左到右，按照步长为s进行选取,例如 print(a[0: 5: 2])输出的结果是[0, 2, 4]
i的值在默认情况下是0，i总是被包含在输出结果中，例如a[0:5] == a[ :5]
j的值在默认情况下是len(a)，也就是列表的长度(5)。j的值不包含在输出结果中，因此a[0: ] 即a[0: 5]可用于输出列表所有元素，因为列表的最大的索引等于列表长度-1。
步长s默认是1，也就是在不指定s的值的情况下，s=1。这也符合我们日常的用法。

当s为负数时：把列表从右往左按照s进行逆序输出。由于方向改变，i的值默认为 -1，也就是最后一个元素；j默认为 -len(a)。

因此采用 a[: : -1]的形式实际上是从列表的最后一个元素向左按照步长s进行选取，也就是逆向输出列表中的每个元素。

回到本例中，我们需要生成一个元素顺序颠倒的列表，因此我们还要在外面套一个中括号（list comprehension）用于生成列表。

Erwarteter Output:
    
    '['ottO', 'leveL', 'leseeseL', [3, 2, 1], 'kajaK', 'relleuM', 'retteR', 'rotoR', [7, 6, 5], 'effE', 'reieM']'





#### b.)
#### (1 Punkte)

Erzeugen Sie eine dritte Liste *thisList3*, die nur diejenigen Einträge aus *thisList* übernimmt, die den Datentype `str` aufweisen.

In [3]:
### Ergänzen Sie hier Ihren Code, um die gewünschte Ausgabe zu erzielen. ###
thisList3 = []

for item in thisList:
    if type(item) == str:
        thisList3.append(item)

### Führen Sie anschließend nachfolgenden Programmcode aus, um Ihre Lösung zu überprüfen ###

print(thisList3)

## Richtig: 1 Punkt

['Otto', 'Level', 'Leseesel', 'Kajak', 'Mueller', 'Retter', 'Rotor', 'Effe', 'Meier']


In [10]:
thisList3 = []

for item in thisList:
    if type(item) == str:
        thisList3.append(item)

print(thisList3)

['Otto', 'Level', 'Leseesel', 'Kajak', 'Mueller', 'Retter', 'Rotor', 'Effe', 'Meier']


函数	说明
type()	返回数据结构类型（list、dict、numpy.ndarray 等）
dtype()	
返回数据元素的数据类型（int、float等）

备注：1）由于 list、dict 等可以包含不同的数据类型，因此不可调用dtype()函数

2）np.array 中要求所有元素属于同一数据类型，因此可调用dtype()函数

astype()	
改变np.array中所有数据元素的数据类型。

备注：能用dtype() 才能用 astype()

Erwarteter Output:
    
    '['Otto', 'Level', 'Leseesel', 'Kajak', 'Mueller', 'Retter', 'Rotor', 'Effe', 'Meier']'

#### c.)
#### (2 Punkte)

In der Liste *evaluations* finden Sie die bereits bekannten Einträge aus *thisList3* zusammen mit weiteren Werten in Tupeln. Sortieren Sie die Einträge von *evaluations* absteigend anhand der **Summe** der Zahleneinträge eines jeden Tupels mithilfe einer `lambda`-Funktion. Geben Sie anschließend das Tupel mit der höchsten Summe aus.

In [4]:
#import random
evaluations = [('Otto', 3.779, 1.238, 0.49), ('Level', 3.961, 5.725, 0.233), ('Leseesel', 3.935, 1.482, 2.41), ('Kajak', 1.989, 2.66, 0.656), ('Mueller', 2.024, 1.427, 3.013), ('Retter', 3.297, 2.36, 3.179), ('Rotor', 1.733, 4.218, 4.972), ('Effe', 3.311, 3.197, 3.991), ('Meier', 5.956, 2.554, 4.622)]

### Bitte ergänzen Sie hier Ihren Code ###

eva_desc = sorted(evaluations, key = lambda x: sum(x[1: ]), reverse = True)
print(eva_desc[0])

## Richtig: 2 Punkte

('Meier', 5.956, 2.554, 4.622)


In [13]:
evaluations = [('Otto', 3.779, 1.238, 0.49), ('Level', 3.961, 5.725, 0.233), ('Leseesel', 3.935, 1.482, 2.41), ('Kajak', 1.989, 2.66, 0.656), ('Mueller', 2.024, 1.427, 3.013), ('Retter', 3.297, 2.36, 3.179), ('Rotor', 1.733, 4.218, 4.972), ('Effe', 3.311, 3.197, 3.991), ('Meier', 5.956, 2.554, 4.622)]

sortEva = sorted(evaluations, key = lambda x: sum(x[1: ]), reverse=True)

print(sortEva)
print(sortEva[0])

[('Meier', 5.956, 2.554, 4.622), ('Rotor', 1.733, 4.218, 4.972), ('Effe', 3.311, 3.197, 3.991), ('Level', 3.961, 5.725, 0.233), ('Retter', 3.297, 2.36, 3.179), ('Leseesel', 3.935, 1.482, 2.41), ('Mueller', 2.024, 1.427, 3.013), ('Otto', 3.779, 1.238, 0.49), ('Kajak', 1.989, 2.66, 0.656)]
('Meier', 5.956, 2.554, 4.622)


Erwarteter Output:

    ('Meier', 5.956, 2.554, 4.622)

#### d.)
#### (1 Punkte)

Oh nein! Leider wurde die Liste leicht beschädigt, sodass nun zusätzliche Listen innerhalb der eigentlichen Listen entstanden sind. Bitte führen Sie die Liste *thisList3Corrupted* mit einem **Python-Ausdruck** wieder in den Ursprungszustand zurück.

In [5]:
thisList3Corrupted = ['Otto', ['Level', 'Leseesel', 'Kajak'], ['Mueller'], 'Retter', 'Rotor', ['Effe', 'Meier']]

### Bitte ergänzen Sie hier Ihren Code ###
thisList3Repaired = []
for item in thisList3Corrupted:
    if type(item) == str:
        thisList3Repaired.append(item)
    else: 
        for element in item:
            thisList3Repaired.append(element)

### Führen Sie anschließend nachfolgenden Programmcode aus, um Ihre Lösung zu überprüfen ###

print(thisList3Repaired)
## Richtig: 1 Punkt

['Otto', 'Level', 'Leseesel', 'Kajak', 'Mueller', 'Retter', 'Rotor', 'Effe', 'Meier']


In [14]:
thisList3Corrupted = ['Otto', ['Level', 'Leseesel', 'Kajak'], ['Mueller'], 'Retter', 'Rotor', ['Effe', 'Meier']]

thisList3Repaired = []

for item in thisList3Corrupted:
    if type(item) == str:
        thisList3Repaired.append(item)
    else:
        for i in item:
            thisList3Repaired.append(i)

print(thisList3Repaired) 

['Otto', 'Level', 'Leseesel', 'Kajak', 'Mueller', 'Retter', 'Rotor', 'Effe', 'Meier']


In [18]:
thisList3Corrupted = ['Otto', ['Level', 'Leseesel', 'Kajak'], ['Mueller'], 'Retter', 'Rotor', ['Effe', 'Meier']]

thisList3Repaired = []

for item in thisList3Corrupted:
    if type(item) == str:
        thisList3Repaired.append(item)
    else:
        thisList3Repaired.extend(item)

print(thisList3Repaired) 

['Otto', 'Level', 'Leseesel', 'Kajak', 'Mueller', 'Retter', 'Rotor', 'Effe', 'Meier']


Erwarteter Output:

    ['Otto', 'Level', 'Leseesel', 'Kajak', 'Mueller', 'Retter', 'Rotor', 'Effe', 'Meier']

### Aufgabe 2 - Einfache Funktionen (7 Punkte)

#### a.)
#### (4 Punkte)

Sie haben gemerkt, dass manche Einträge der Liste *thisList* eine besondere Eigenschaft besitzen. Es handelt sich bei einigen Einträgen um Palindrome, die rückwärts gelesen genau denselben Text ergeben, wie in der eigentlichen Leserichtung.

Um nun möglichst schnell herauszufinden, welche Einträge einer Liste Palindrome darstellen, schreiben Sie bitte eine Funktion `CheckForPalindroms`, der eine **Liste** übergeben wird. Als Rückgabewert soll die Funktion anschließend in einer Liste die Indices der Elemente zurückgeben, die als Palindrom identifiziert wurden.

**Tipp:**
<details>

Achten Sie darauf, dass gilt: 'a' != 'A'. Finden Sie also zunächst eine Funktion, um die Strings zu formatieren.
</details>

In [6]:
thisList = ['Otto', 'Level', 'Leseesel', [1, 2, 3], 'Kajak', 'Mueller', 'Retter', 'Rotor', [5, 6, 7], 'Effe', 'Meier']

### Bitte ergänzen Sie hier Ihren Programmcode ###
def CheckForPalindroms(list):
    ind_list = []
    for i in range(len(list)):
        if type(list[i]) == str:
            low=list[i].lower()
            
            if low[::-1] == low:
                ind_list.append(i)
    return ind_list

        


### Führen Sie anschließend nachfolgenden Programmcode aus, um Ihre Lösung zu überprüfen ###

CheckForPalindroms(thisList)

## Richtig: 4 Punkte

[0, 1, 2, 4, 6, 7, 9]

In [19]:
thisList = ['Otto', 'Level', 'Leseesel', [1, 2, 3], 'Kajak', 'Mueller', 'Retter', 'Rotor', [5, 6, 7], 'Effe', 'Meier']

def CheckForPalindroms(list):
    index_list = []
    for i in range(len(list)):
        if type(list[i]) == str:
            element = list[i].upper()
            if element == element[::-1]:
                index_list.append(i)
        else:
            if list[i] == list[i][::-1]:
                index_list.append(i)

    return index_list

CheckForPalindroms(thisList)

[0, 1, 2, 4, 6, 7, 9]

Erwarteter Output:

    
    [0, 1, 2, 4, 6, 7, 9]








#### b.)
#### (3 Punkte)

Schreiben Sie eine Funktion ``splitNumber``, die eine ihr übergebene Gleitkommazahl in deren ganzzahligen Teil und den Dezimal-Teil aufteilt und beides zusammen als Tupel ausgibt. Nutzen Sie dafür **ausschließlich die Ihnen bekannten Operatoren** und **keine built-in-Funktionen**.

Für den Fall, dass keine Zahl übergeben wird, soll eine `Exception` als `TypeError` ausgegeben werden. Informieren Sie sich dazu zum Konstrukt *try-except* und wenden dieses an.

In [11]:
### Bitte ergänzen Sie hier Ihren Programmcode ###
def splitNumber(num):
    try:
        num_ganz = int(num)
        num_dez = num-num_ganz
        return (num_ganz, num_dez)


    except ValueError as e:
        print(e)
        print("TypeError: Input has wrong type!")

### Führen Sie anschließend nachfolgenden Programmcode aus, um Ihre Lösung zu überprüfen ###

print(splitNumber(3.56))
print(splitNumber("Hallo"))

## Richtig: 2,5 Punkte
## 0,5 Punkte Abzug, da ValueError statt TypeError genutzt

(3, 0.56)
invalid literal for int() with base 10: 'Hallo'
TypeError: Input has wrong type!
None


In [25]:
def splitNumber(num):
    try:
        int_num = int(num)
        dec_num = num - int_num
        return(int_num, dec_num)

    except ValueError :
        print('TypeError: Input has wrong type!')

print(splitNumber(3.56))
print(splitNumber('Hello'))

(3, 0.56)
TypeError: Input has wrong type!
None


Erwarteter Output:

    (3, 0.56)
    
    ...(Hier steht eine Fehlerbeschreibung)...

    TypeError: Input has wrong type!

### Aufgabe 3 - Klassendefinition & Weitere Funktionen (12 Punkte)

#### a.) Die Rechteck-Klasse
#### (8 Punkte)

Erstellen Sie eine Klasse *Rechteck*, die als Parameter eine Id *id*, eine Breite *width* und eine Höhe *height* entgegennimmt. Die Methode `DetermineArea()` soll die Fläche des Rechtecks berechnen. Diese Methode soll direkt in der `__init__()`-Methode aufgerufen werden. Zusätzlich soll die `__str__()`-Methode für eine entsprechende Ausgabe sorgen.

Anschließend definieren Sie eine Funktion `Load` (diese ist nicht Teil der Klasse *Rechteck*), welche die Datei `Rectangulars.json` einliest, aus den enthaltenen Informationen Objekte der Klasse *Rechteck* erzeugt und alle Objekte in einer Liste ausgibt.

In [26]:
import json
### Bitte ergänzen Sie hier Ihren Programmcode ###
class Rechteck():

    def __init__(self, id, width, height):
        self.id = id
        self.width = width
        self.height = height
        self.area = self.DetermineArea()

    def DetermineArea(self):
        return self.width * self.height

    def __str__(self):
        return f'Recteck {self.id} mit Breite {self.width}, Höhe {self.height}, Fläche {self.area}'
    
def Load(name):
    rec_list = []
    with open(name) as file:
        data = json.load(file)
        for id in data.keys():
            rec = Rechteck(id, data[id]["width"], data[id]["height"])
            rec_list.append(rec)

    return rec_list



### Führen Sie anschließend nachfolgenden Programmcode aus, um Ihre Lösung zu überprüfen ###

rectangulars = Load('Rectangulars.json')

# for rec in rectangulars:
#     print(rec)
print(rectangulars)  # 这个print作用于由对象组成的列表而不是对象，因此str函数不会调用
## Richtig: 8 Punkte

[<__main__.Rechteck object at 0x000001B2F4F798B0>, <__main__.Rechteck object at 0x000001B2F4F79730>, <__main__.Rechteck object at 0x000001B2F4F79AC0>, <__main__.Rechteck object at 0x000001B2F4F79A00>, <__main__.Rechteck object at 0x000001B2F4F79520>, <__main__.Rechteck object at 0x000001B2F4F79940>, <__main__.Rechteck object at 0x000001B2F4F799A0>, <__main__.Rechteck object at 0x000001B2F4F79FA0>, <__main__.Rechteck object at 0x000001B2F4F79F40>, <__main__.Rechteck object at 0x000001B2F4F79E80>]


In [30]:
import json

class Rectangle:
    def __init__(self, id, width, height):
        # 小写字母来自参数表，大写字母才是类的属性！！
        # 想知道构造方法的参数表，可以看一下json文件，也就是我们手头有什么数据可以用于构造对象
        self.ID = id
        self.Width = width
        self.Height = height

        # 注意：面积虽然也是属性，但是我们不能把面积写到类的构造方法中去！！
        self.Area = self.DetermineArea()

    def DetermineArea(self):
        area = self.Width * self.Height
        return area
        # return self.Width * self.Height

    def __str__(self):
        return f'Rectangle {self.ID} with width {self.Width}, height {self.Height}, area {self.Area}'
        # 注意这里不能用print，不然会输出None

    
def Load(path):
    recList = []

    with open(path) as file:
        data = json.load(file) # data 的类型是字典

        for id in data.keys():
            rec = Rectangle(id, data[id]['width'], data[id]['height'])
            recList.append(rec)

    return recList

# 从文件中读入数据
rectangles = Load("Rectangulars.json")  # rectangles 的类型是list，元素是长方形对象

# 打印由对象组成的列表，输出一个由对象地址组成的列表
print(rectangles)

# 遍历对象列表并打印，这样才能调用str函数，输出我们想要的信息
for rec in rectangles:
    print(rec)




        

[<__main__.Rectangle object at 0x000001B2F4F4AC10>, <__main__.Rectangle object at 0x000001B2F50F7BE0>, <__main__.Rectangle object at 0x000001B2F4F3C310>, <__main__.Rectangle object at 0x000001B2F4F3C6D0>, <__main__.Rectangle object at 0x000001B2F4F80130>, <__main__.Rectangle object at 0x000001B2F4F801C0>, <__main__.Rectangle object at 0x000001B2F4F80E50>, <__main__.Rectangle object at 0x000001B2F4F809A0>, <__main__.Rectangle object at 0x000001B2F4F80880>, <__main__.Rectangle object at 0x000001B2F4F80FA0>]
Rectangle 0 with width 6, height 6, area 36
Rectangle 1 with width 7, height 10, area 70
Rectangle 2 with width 9, height 6, area 54
Rectangle 3 with width 7, height 10, area 70
Rectangle 4 with width 5, height 11, area 55
Rectangle 5 with width 10, height 10, area 100
Rectangle 6 with width 10, height 13, area 130
Rectangle 7 with width 8, height 6, area 48
Rectangle 8 with width 8, height 7, area 56
Rectangle 9 with width 9, height 10, area 90


Erwarteter Output:

    Rechteck 0 mit Breite 6, Höhe 6, Fläche 36
    Rechteck 1 mit Breite 7, Höhe 10, Fläche 70
    Rechteck 2 mit Breite 9, Höhe 6, Fläche 54
    Rechteck 3 mit Breite 7, Höhe 10, Fläche 70
    Rechteck 4 mit Breite 5, Höhe 11, Fläche 55
    Rechteck 5 mit Breite 10, Höhe 10, Fläche 100
    Rechteck 6 mit Breite 10, Höhe 13, Fläche 130
    Rechteck 7 mit Breite 8, Höhe 6, Fläche 48
    Rechteck 8 mit Breite 8, Höhe 7, Fläche 56
    Rechteck 9 mit Breite 9, Höhe 10, Fläche 90

#### b.) Eine letzte Nachricht
#### (4 Punkte)

Ihnen wurde eine mysteriöse Nachricht zugeschickt. Zusätzlich zu dieser Nachricht standen auf einem weiteren Zettel nur die Worte "Veni, vidi, vici." und der Hinweis "Schlüssel ist die 20".

Schreiben Sie eine Funktion `Decrypt`, die die verschlüsselte Nachricht decodieren kann. Achten Sie dabei auf Leerzeichen sowie Sonderzeichen.

**Tipp 1:**
<details>

Informieren Sie sich über die Cäsar-Verschlüsselung und deren Funktionsweise!
</details>

**Tipp 2:**
<details>

Informieren Sie sich über die Funktion ``ord(x)`` und achten Sie auf Groß- und Kleinschreibung. Des Weiteren ist auch die Funktion ``.isalpha()`` hilfreich. Letztere lässt sich auf Strings anwenden. 
</details>

In [13]:
verschluesselteNachricht = "Ylmny Buomuozauvy aymwbuzzn! Scjjcy :)"
schluessel = 20

### Bitte ergänzen Sie hier Ihren Programmcode ###
def Decrypt(text, key):
    klartext = ''
    for sign in text:
        if sign >= 'A' and sign <= "Z":
            klartext += chr(ord("A")+((ord(sign)-ord("A"))-key) % 26)

        elif sign >= 'a' and sign <= "z":
            klartext += chr(ord("a")+((ord(sign)-ord("a"))-key) % 26)

        else:
            klartext += sign
    return klartext
            



### Führen Sie anschließend nachfolgenden Programmcode aus, um Ihre Lösung zu überprüfen ###

Decrypt(verschluesselteNachricht, schluessel)

## Richtig: 4 Punkte

'Erste Hausaufgabe geschafft! Yippie :)'

In [31]:
ciphertext = "Ylmny Buomuozauvy aymwbuzzn! Scjjcy :)"
key = 20

def Decrypt(text, key):
    message = '' # 这里是明文，初始化为空字符。python中的str类型可以用加号连接也可以遍历

    for sign in text:
        if sign >= 'A' and sign <= 'Z':  # 判断sign是否位于大写字母A到Z之间
            message += chr(ord('A') + ((ord(sign)-ord("A"))-key) % 26)

        elif sign >= 'a' and sign <= "z":
            message += chr(ord("a")+((ord(sign)-ord("a"))-key) % 26)

        else: # sign是特殊符号的情况
            message += sign

    return message

Decrypt(ciphertext, key)


'Erste Hausaufgabe geschafft! Yippie :)'